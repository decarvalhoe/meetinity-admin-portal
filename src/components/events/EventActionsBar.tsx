import React, { useMemo, useState } from 'react'

interface EventActionsBarProps {
  selected: string[]
  onApprove: () => void
  onReject: () => void
  onArchive: () => void
  onCreate: () => void
  onManageCategories: () => void
  onRefresh: () => void
  onApplyTags: (tags: string[]) => void
  availableTags: string[]
}

export function EventActionsBar({
  selected,
  onApprove,
  onReject,
  onArchive,
  onCreate,
  onManageCategories,
  onRefresh,
  onApplyTags,
  availableTags
}: EventActionsBarProps) {
  const [tagsInput, setTagsInput] = useState('')

  const isSelectionEmpty = selected.length === 0
  const parsedTags = useMemo(
    () =>
      tagsInput
        .split(',')
        .map(tag => tag.trim())
        .filter(Boolean),
    [tagsInput]
  )

  const canApplyTags = parsedTags.length > 0 && !isSelectionEmpty

  const handleApplyTags = () => {
    if (parsedTags.length === 0) {
      return
    }
    onApplyTags(parsedTags)
    setTagsInput('')
  }

  return (
    <section className="event-actions">
      <div className="event-actions__primary">
        <button type="button" onClick={onCreate}>
          Créer un événement
        </button>
        <button type="button" onClick={onManageCategories}>
          Gérer les catégories
        </button>
        <button type="button" onClick={onRefresh}>
          Rafraîchir
        </button>
      </div>
      <div className="event-actions__bulk">
        <label className="event-actions__tags-field">
          <span className="event-actions__label">Tags</span>
          <input
            placeholder="tag1, tag2"
            value={tagsInput}
            onChange={event => setTagsInput(event.target.value)}
            list="event-tags-suggestions"
          />
          <datalist id="event-tags-suggestions">
            {availableTags.map(tag => (
              <option key={tag} value={tag} />
            ))}
          </datalist>
        </label>
        <button type="button" disabled={!canApplyTags} onClick={handleApplyTags}>
          Appliquer les tags
        </button>
        <button type="button" disabled={isSelectionEmpty} onClick={onApprove}>
          Approuver
        </button>
        <button type="button" disabled={isSelectionEmpty} onClick={onReject}>
          Rejeter
        </button>
        <button type="button" disabled={isSelectionEmpty} onClick={onArchive}>
          Archiver
        </button>
      </div>
    </section>
  )
}
