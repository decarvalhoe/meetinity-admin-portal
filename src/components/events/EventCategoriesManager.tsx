import React, { useState } from 'react'
import { EventService, type EventCategory } from '../../services/eventService'

interface EventCategoriesManagerProps {
  categories: EventCategory[]
  onCategoriesChange: () => Promise<void> | void
}

export function EventCategoriesManager({ categories, onCategoriesChange }: EventCategoriesManagerProps) {
  const [name, setName] = useState('')
  const [isSubmitting, setIsSubmitting] = useState(false)
  const [editingId, setEditingId] = useState<string | null>(null)
  const [editingName, setEditingName] = useState('')

  const handleCreate = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault()
    if (!name.trim()) {
      return
    }
    setIsSubmitting(true)
    try {
      await EventService.createCategory({ name: name.trim() })
      setName('')
      await onCategoriesChange()
    } finally {
      setIsSubmitting(false)
    }
  }

  const handleStartEdit = (category: EventCategory) => {
    setEditingId(category.id)
    setEditingName(category.name)
  }

  const handleSaveEdit = async (categoryId: string) => {
    if (!editingName.trim()) {
      return
    }
    setIsSubmitting(true)
    try {
      await EventService.updateCategory(categoryId, { name: editingName.trim() })
      setEditingId(null)
      setEditingName('')
      await onCategoriesChange()
    } finally {
      setIsSubmitting(false)
    }
  }

  const handleDelete = async (categoryId: string) => {
    setIsSubmitting(true)
    try {
      await EventService.deleteCategory(categoryId)
      await onCategoriesChange()
    } finally {
      setIsSubmitting(false)
    }
  }

  return (
    <section className="event-categories">
      <h3>Gestion des catégories</h3>
      <form className="event-categories__create" onSubmit={handleCreate}>
        <input
          placeholder="Nom de la catégorie"
          value={name}
          onChange={event => setName(event.target.value)}
          disabled={isSubmitting}
        />
        <button type="submit" disabled={isSubmitting || name.trim().length === 0}>
          Ajouter
        </button>
      </form>
      <ul className="event-categories__list">
        {categories.map(category => (
          <li key={category.id} className="event-categories__item">
            {editingId === category.id ? (
              <>
                <input value={editingName} onChange={event => setEditingName(event.target.value)} />
                <button type="button" onClick={() => handleSaveEdit(category.id)} disabled={isSubmitting}>
                  Enregistrer
                </button>
                <button
                  type="button"
                  onClick={() => {
                    setEditingId(null)
                    setEditingName('')
                  }}
                  disabled={isSubmitting}
                >
                  Annuler
                </button>
              </>
            ) : (
              <>
                <span className="event-categories__name">{category.name}</span>
                <button type="button" onClick={() => handleStartEdit(category)} disabled={isSubmitting}>
                  Renommer
                </button>
                <button type="button" onClick={() => handleDelete(category.id)} disabled={isSubmitting}>
                  Supprimer
                </button>
              </>
            )}
          </li>
        ))}
        {categories.length === 0 && <li>Aucune catégorie disponible</li>}
      </ul>
    </section>
  )
}
