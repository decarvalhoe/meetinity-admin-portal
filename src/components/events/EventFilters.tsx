import React from 'react'
import type { EventCategory } from '../../services/eventService'

export interface EventFiltersState {
  search: string
  status: string
  categoryId: string
  startDate: string
  endDate: string
}

interface EventFiltersProps {
  filters: EventFiltersState
  onChange: (filters: Partial<EventFiltersState>) => void
  categories: EventCategory[]
  statuses?: Array<{ value: string; label: string }>
  disableStatus?: boolean
}

export function EventFilters({ filters, onChange, categories, statuses, disableStatus }: EventFiltersProps) {
  const statusOptions =
    statuses ?? (
      [
        { value: '', label: 'Tous les statuts' },
        { value: 'draft', label: 'Brouillons' },
        { value: 'pending', label: 'En attente' },
        { value: 'published', label: 'Publiés' },
        { value: 'rejected', label: 'Rejetés' },
        { value: 'archived', label: 'Archivés' }
      ] as Array<{ value: string; label: string }>
    )

  return (
    <section className="event-filters">
      <div className="event-filters__row">
        <label className="event-filters__field">
          <span className="event-filters__label">Recherche</span>
          <input
            placeholder="Rechercher un événement"
            value={filters.search}
            onChange={e => onChange({ search: e.target.value })}
          />
        </label>
        <label className="event-filters__field">
          <span className="event-filters__label">Statut</span>
          <select
            value={filters.status}
            disabled={disableStatus}
            onChange={e => onChange({ status: e.target.value })}
          >
            {statusOptions.map(option => (
              <option key={option.value} value={option.value}>
                {option.label}
              </option>
            ))}
          </select>
        </label>
        <label className="event-filters__field">
          <span className="event-filters__label">Catégorie</span>
          <select value={filters.categoryId} onChange={e => onChange({ categoryId: e.target.value })}>
            <option value="">Toutes les catégories</option>
            {categories.map(category => (
              <option key={category.id} value={category.id}>
                {category.name}
              </option>
            ))}
          </select>
        </label>
      </div>
      <div className="event-filters__row">
        <label className="event-filters__field">
          <span className="event-filters__label">Date de début</span>
          <input
            type="date"
            value={filters.startDate}
            onChange={e => onChange({ startDate: e.target.value })}
          />
        </label>
        <label className="event-filters__field">
          <span className="event-filters__label">Date de fin</span>
          <input type="date" value={filters.endDate} onChange={e => onChange({ endDate: e.target.value })} />
        </label>
      </div>
    </section>
  )
}
