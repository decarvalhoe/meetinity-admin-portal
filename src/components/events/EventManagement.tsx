import React, { useCallback, useEffect, useMemo, useReducer, useRef, useState } from 'react'
import {
  Event,
  EventCategory,
  EventInput,
  EventService,
  EventAnalyticsSummary,
  EventAttendanceSeries,
  EventConversionStat,
  EventApprovalStage,
  EventEngagementCell
} from '../../services/eventService'
import { EventFilters, EventFiltersState } from './EventFilters'
import { EventTable } from './EventTable'
import { EventActionsBar } from './EventActionsBar'
import { EventForm } from './EventForm'
import { EventCategoriesManager } from './EventCategoriesManager'
import { useDebounce } from '../../hooks/useDebounce'
import { usePagination } from '../../hooks/usePagination'
import { useWebSocket } from '../../hooks/useWebSocket'
import { EventAnalyticsDashboard, AnalyticsRangeOption } from './EventAnalyticsDashboard'

export interface EventAnalyticsState {
  summary: EventAnalyticsSummary | null
  attendance: EventAttendanceSeries[]
  conversions: EventConversionStat[]
  funnel: EventApprovalStage[]
  heatmap: EventEngagementCell[]
  updatedAt?: string
}

const defaultSummary: EventAnalyticsSummary = {
  totalEvents: 0,
  published: 0,
  pendingApproval: 0,
  rejected: 0,
  averageAttendanceRate: 0,
  conversionRate: 0
}

const initialAnalyticsState: EventAnalyticsState = {
  summary: null,
  attendance: [],
  conversions: [],
  funnel: [],
  heatmap: [],
  updatedAt: undefined
}

type AnalyticsAction =
  | {
      type: 'SET_BASELINE'
      payload: {
        summary: EventAnalyticsSummary
        attendance: EventAttendanceSeries[]
        conversions: EventConversionStat[]
        funnel: EventApprovalStage[]
        heatmap: EventEngagementCell[]
        updatedAt?: string
      }
    }
  | { type: 'PATCH_SUMMARY'; payload: Partial<EventAnalyticsSummary> }
  | {
      type: 'UPSERT_ATTENDANCE_POINT'
      payload: {
        eventId: string
        eventName: string
        date: string
        registrations: number
        attendance: number
      }
    }
  | { type: 'UPSERT_CONVERSION'; payload: EventConversionStat }
  | { type: 'UPSERT_FUNNEL_STAGE'; payload: EventApprovalStage }
  | { type: 'UPSERT_HEATMAP_CELL'; payload: EventEngagementCell }

export function analyticsReducer(state: EventAnalyticsState, action: AnalyticsAction): EventAnalyticsState {
  switch (action.type) {
    case 'SET_BASELINE':
      return {
        summary: action.payload.summary,
        attendance: action.payload.attendance,
        conversions: action.payload.conversions,
        funnel: action.payload.funnel,
        heatmap: action.payload.heatmap,
        updatedAt: action.payload.updatedAt ?? new Date().toISOString()
      }
    case 'PATCH_SUMMARY': {
      const nextSummary = { ...(state.summary ?? defaultSummary), ...action.payload }
      return {
        ...state,
        summary: nextSummary,
        updatedAt: new Date().toISOString()
      }
    }
    case 'UPSERT_ATTENDANCE_POINT': {
      const { eventId, eventName, date, registrations, attendance } = action.payload
      const nextAttendance = state.attendance.map(series => ({ ...series, series: [...series.series] }))
      const existingSeriesIndex = nextAttendance.findIndex(series => series.eventId === eventId)

      if (existingSeriesIndex === -1) {
        nextAttendance.push({
          eventId,
          eventName,
          series: [{ date, registrations, attendance }]
        })
      } else {
        const existingSeries = nextAttendance[existingSeriesIndex]
        const points = [...existingSeries.series]
        const pointIndex = points.findIndex(point => point.date === date)
        const nextPoint = { date, registrations, attendance }

        if (pointIndex === -1) {
          points.push(nextPoint)
        } else {
          points[pointIndex] = nextPoint
        }

        points.sort((a, b) => new Date(a.date).getTime() - new Date(b.date).getTime())
        nextAttendance[existingSeriesIndex] = { ...existingSeries, eventName, series: points }
      }

      return {
        ...state,
        attendance: nextAttendance,
        updatedAt: new Date().toISOString()
      }
    }
    case 'UPSERT_CONVERSION': {
      const index = state.conversions.findIndex(item => item.stage === action.payload.stage)
      const nextConversions = [...state.conversions]
      if (index === -1) {
        nextConversions.push(action.payload)
      } else {
        nextConversions[index] = action.payload
      }
      return {
        ...state,
        conversions: nextConversions,
        updatedAt: new Date().toISOString()
      }
    }
    case 'UPSERT_FUNNEL_STAGE': {
      const index = state.funnel.findIndex(stage => stage.stage === action.payload.stage)
      const nextFunnel = [...state.funnel]
      if (index === -1) {
        nextFunnel.push(action.payload)
      } else {
        nextFunnel[index] = action.payload
      }
      return {
        ...state,
        funnel: nextFunnel,
        updatedAt: new Date().toISOString()
      }
    }
    case 'UPSERT_HEATMAP_CELL': {
      const index = state.heatmap.findIndex(
        cell => cell.day === action.payload.day && cell.hour === action.payload.hour
      )
      const nextHeatmap = [...state.heatmap]
      if (index === -1) {
        nextHeatmap.push(action.payload)
      } else {
        nextHeatmap[index] = action.payload
      }
      return {
        ...state,
        heatmap: nextHeatmap,
        updatedAt: new Date().toISOString()
      }
    }
    default:
      return state
  }
}

interface EventManagementProps {
  initialFilters?: Partial<EventFiltersState>
  lockedStatus?: string
}

export function EventManagement({ initialFilters, lockedStatus }: EventManagementProps) {
  const initialState = useMemo<EventFiltersState>(
    () => ({
      search: initialFilters?.search ?? '',
      status: lockedStatus ?? initialFilters?.status ?? '',
      categoryId: initialFilters?.categoryId ?? '',
      startDate: initialFilters?.startDate ?? '',
      endDate: initialFilters?.endDate ?? ''
    }),
    [initialFilters, lockedStatus]
  )

  const [events, setEvents] = useState<Event[]>([])
  const [total, setTotal] = useState(0)
  const [filters, setFilters] = useState<EventFiltersState>(initialState)
  const [categories, setCategories] = useState<EventCategory[]>([])
  const [availableTags, setAvailableTags] = useState<string[]>([])
  const [selected, setSelected] = useState<string[]>([])
  const [selectionResetKey, setSelectionResetKey] = useState(0)
  const [editingEvent, setEditingEvent] = useState<Event | null>(null)
  const [isFormOpen, setIsFormOpen] = useState(false)
  const [isCategoriesOpen, setIsCategoriesOpen] = useState(false)
  const [analyticsRange, setAnalyticsRange] = useState<AnalyticsRangeOption>('30d')
  const [analyticsLoading, setAnalyticsLoading] = useState(false)

  const [analyticsState, dispatchAnalytics] = useReducer(analyticsReducer, initialAnalyticsState)

  const debouncedSearch = useDebounce(filters.search)
  const { page, pageSize, setPage } = usePagination(20)

  const requestIdRef = useRef(0)

  useEffect(() => {
    if (lockedStatus && filters.status !== lockedStatus) {
      setFilters(prev => ({ ...prev, status: lockedStatus }))
    }
  }, [filters.status, lockedStatus])

  const loadEvents = useCallback(async () => {
    const requestId = requestIdRef.current + 1
    requestIdRef.current = requestId

    const res = await EventService.list({
      page,
      pageSize,
      search: debouncedSearch,
      status: lockedStatus ?? filters.status,
      categoryId: filters.categoryId,
      startDate: filters.startDate,
      endDate: filters.endDate
    })

    if (requestId !== requestIdRef.current) {
      return null
    }

    setEvents(res.events)
    setTotal(res.total)
    return res
  }, [
    page,
    pageSize,
    debouncedSearch,
    filters.status,
    filters.categoryId,
    filters.startDate,
    filters.endDate,
    lockedStatus
  ])

  const ensurePageWithinBounds = useCallback(
    (count: number) => {
      const maxPage = Math.max(0, Math.ceil(count / pageSize) - 1)
      if (page > maxPage) {
        setPage(maxPage)
      }
    },
    [page, pageSize, setPage]
  )

  const loadCategories = useCallback(async () => {
    const data = await EventService.listCategories()
    setCategories(data)
  }, [])

  const loadTags = useCallback(async () => {
    try {
      const data = await EventService.listTags()
      setAvailableTags(data)
    } catch (error) {
      console.error('Failed to load event tags', error)
    }
  }, [])

  useEffect(() => {
    loadCategories()
    loadTags()
  }, [loadCategories, loadTags])

  const loadAnalytics = useCallback(async (range: AnalyticsRangeOption) => {
    setAnalyticsLoading(true)
    try {
      const [summary, attendanceSeries, conversionSeries, funnelSeries, heatmapSeries] =
        await Promise.all([
          EventService.getAnalyticsSummary({ range }),
          EventService.getAttendanceAnalytics({ range }),
          EventService.getConversionAnalytics({ range }),
          EventService.getApprovalFunnel({ range }),
          EventService.getEngagementHeatmap({ range })
        ])

      dispatchAnalytics({
        type: 'SET_BASELINE',
        payload: {
          summary,
          attendance: attendanceSeries,
          conversions: conversionSeries,
          funnel: funnelSeries,
          heatmap: heatmapSeries,
          updatedAt: new Date().toISOString()
        }
      })
    } catch (error) {
      console.error('Failed to load event analytics', error)
    } finally {
      setAnalyticsLoading(false)
    }
  }, [])

  useEffect(() => {
    loadAnalytics(analyticsRange)
  }, [analyticsRange, loadAnalytics])

  const analyticsSocketUrl = useMemo(() => {
    const wsBase = (import.meta.env.VITE_WS_BASE_URL as string | undefined)?.replace(/\/$/, '')
    if (wsBase) {
      return `${wsBase}/events`
    }
    const apiBase = (import.meta.env.VITE_API_BASE_URL as string | undefined)?.replace(/\/$/, '')
    if (!apiBase) {
      return null
    }
    if (apiBase.startsWith('http')) {
      return `${apiBase.replace(/^http/, 'ws')}/ws/events`
    }
    return `ws://${apiBase}/ws/events`
  }, [])

  const { readyState: analyticsSocketState, subscribe: subscribeToAnalyticsStream } = useWebSocket(
    analyticsSocketUrl,
    {
      enabled: Boolean(analyticsSocketUrl),
      reconnectInterval: 4000,
      maxReconnectInterval: 30000
    }
  )

  useEffect(() => {
    if (!analyticsSocketUrl) {
      return undefined
    }
    return subscribeToAnalyticsStream(event => {
      try {
        const message = JSON.parse(event.data as string)
        if (!message || typeof message !== 'object') {
          return
        }
        switch (message.type) {
          case 'summary':
            dispatchAnalytics({ type: 'PATCH_SUMMARY', payload: message.data ?? {} })
            break
          case 'registration':
            if (!message.eventId || !message.date) {
              break
            }
            dispatchAnalytics({
              type: 'UPSERT_ATTENDANCE_POINT',
              payload: {
                eventId: message.eventId,
                eventName: message.eventName,
                date: message.date,
                registrations: message.registrations,
                attendance: message.attendance
              }
            })
            break
          case 'conversion':
            if (!message.stage) {
              break
            }
            dispatchAnalytics({
              type: 'UPSERT_CONVERSION',
              payload: { stage: message.stage, value: message.value }
            })
            break
          case 'approval':
            if (!message.stage) {
              break
            }
            dispatchAnalytics({
              type: 'UPSERT_FUNNEL_STAGE',
              payload: { stage: message.stage, count: message.count }
            })
            break
          case 'heatmap':
            if (!message.day || typeof message.hour !== 'number') {
              break
            }
            dispatchAnalytics({
              type: 'UPSERT_HEATMAP_CELL',
              payload: { day: message.day, hour: message.hour, value: message.value }
            })
            break
          default:
            console.warn('Unhandled analytics event', message)
        }
      } catch (error) {
        console.error('Failed to parse analytics update', error)
      }
    })
  }, [subscribeToAnalyticsStream])

  const handleAnalyticsRangeChange = (range: AnalyticsRangeOption) => {
    setAnalyticsRange(range)
  }

  useEffect(() => {
    const refresh = async () => {
      const res = await loadEvents()
      if (res) {
        ensurePageWithinBounds(res.total)
      }
    }

    refresh()
  }, [loadEvents, ensurePageWithinBounds])

  const handleFilterChange = (partial: Partial<EventFiltersState>) => {
    setFilters(prev => {
      const next = { ...prev, ...partial }
      if (lockedStatus) {
        next.status = lockedStatus
      }
      return next
    })
    setPage(0)
  }

  const handleBulkAction = async (action: 'approve' | 'reject' | 'archive') => {
    if (selected.length === 0) {
      return
    }
    await EventService.bulk(selected, action)
    setSelected([])
    setSelectionResetKey(prev => prev + 1)
    const res = await loadEvents()
    if (res) {
      ensurePageWithinBounds(res.total)
    }
  }

  const handleApplyTags = async (tags: string[]) => {
    if (selected.length === 0 || tags.length === 0) {
      return
    }
    await EventService.bulkTags(selected, tags)
    setSelected([])
    setSelectionResetKey(prev => prev + 1)
    const res = await loadEvents()
    if (res) {
      ensurePageWithinBounds(res.total)
    }
    await loadTags()
  }

  const handleCreate = () => {
    setEditingEvent(null)
    setIsFormOpen(true)
  }

  const handleEdit = (event: Event) => {
    setEditingEvent(event)
    setIsFormOpen(true)
  }

  const handleSave = async (payload: EventInput) => {
    if (editingEvent) {
      await EventService.update(editingEvent.id, payload)
    } else {
      await EventService.create(payload)
    }
    setIsFormOpen(false)
    setEditingEvent(null)
    const res = await loadEvents()
    if (res) {
      ensurePageWithinBounds(res.total)
    }
    await loadTags()
  }

  const handleCancelForm = () => {
    setIsFormOpen(false)
    setEditingEvent(null)
  }

  const handleToggleCategories = () => {
    setIsCategoriesOpen(open => !open)
  }

  const handleRefresh = async () => {
    const res = await loadEvents()
    if (res) {
      ensurePageWithinBounds(res.total)
    }
    await loadAnalytics(analyticsRange)
  }

  return (
    <div className="event-management">
      <EventAnalyticsDashboard
        summary={analyticsState.summary}
        attendance={analyticsState.attendance}
        conversions={analyticsState.conversions}
        funnel={analyticsState.funnel}
        heatmap={analyticsState.heatmap}
        loading={analyticsLoading}
        lastUpdated={analyticsState.updatedAt}
        connectionState={analyticsSocketState}
        onRangeChange={handleAnalyticsRangeChange}
        range={analyticsRange}
      />
      <EventFilters
        filters={filters}
        onChange={handleFilterChange}
        categories={categories}
        disableStatus={Boolean(lockedStatus)}
      />
      <EventActionsBar
        selected={selected}
        onApprove={() => handleBulkAction('approve')}
        onReject={() => handleBulkAction('reject')}
        onArchive={() => handleBulkAction('archive')}
        onCreate={handleCreate}
        onManageCategories={handleToggleCategories}
        onRefresh={handleRefresh}
        onApplyTags={handleApplyTags}
        availableTags={availableTags}
      />
      {isFormOpen && (
        <EventForm
          event={editingEvent}
          categories={categories}
          availableTags={availableTags}
          onSave={handleSave}
          onCancel={handleCancelForm}
        />
      )}
      {isCategoriesOpen && (
        <EventCategoriesManager categories={categories} onCategoriesChange={loadCategories} />
      )}
      <EventTable
        data={events}
        total={total}
        page={page}
        pageSize={pageSize}
        onPageChange={setPage}
        onSelect={setSelected}
        onEdit={handleEdit}
        clearSelectionKey={selectionResetKey}
      />
    </div>
  )
}
