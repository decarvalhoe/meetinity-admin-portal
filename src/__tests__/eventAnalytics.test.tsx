/**
 * @vitest-environment jsdom
 */
import React from 'react'
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest'
import { render, screen, waitFor, act } from '@testing-library/react'
import { EventManagement, analyticsReducer, EventAnalyticsState } from '../components/events/EventManagement'
import type {
  EventApprovalStage,
  EventAttendanceSeries,
  EventConversionStat,
  EventEngagementCell,
  EventAnalyticsSummary
} from '../services/eventService'

const subscribers = new Set<(event: MessageEvent) => void>()
let emitWebSocketMessage: (payload: unknown) => void
let hasActiveSubscribers = false

emitWebSocketMessage = payload => {
  const message = typeof payload === 'string' ? payload : JSON.stringify(payload)
  subscribers.forEach(handler =>
    handler({
      data: message
    } as MessageEvent)
  )
}

var fixtures: {
  summary: EventAnalyticsSummary
  attendance: EventAttendanceSeries[]
  conversions: EventConversionStat[]
  funnel: EventApprovalStage[]
  heatmap: EventEngagementCell[]
}

vi.mock('../services/eventService', () => {
  fixtures = {
    summary: {
      totalEvents: 42,
      published: 30,
      pendingApproval: 6,
      rejected: 6,
      averageAttendanceRate: 0.72,
      conversionRate: 0.58
    },
    attendance: [
      {
        eventId: 'a',
        eventName: 'Forum Product',
        series: [
          { date: '2024-01-01', registrations: 120, attendance: 110 },
          { date: '2024-01-02', registrations: 140, attendance: 135 }
        ]
      }
    ],
    conversions: [
      { stage: 'Visiteurs', value: 420 },
      { stage: 'Inscrits', value: 190 },
      { stage: 'Présents', value: 140 }
    ],
    funnel: [
      { stage: 'Soumis', count: 40 },
      { stage: 'Vérification', count: 28 },
      { stage: 'Approuvés', count: 22 }
    ],
    heatmap: [
      { day: 'Lundi', hour: 9, value: 30 },
      { day: 'Lundi', hour: 10, value: 50 },
      { day: 'Mardi', hour: 14, value: 70 }
    ]
  }

  return {
    EventService: {
      list: vi.fn().mockResolvedValue({ events: [], total: 0 }),
      get: vi.fn(),
      create: vi.fn(),
      update: vi.fn(),
      remove: vi.fn(),
      approve: vi.fn(),
      reject: vi.fn(),
      archive: vi.fn(),
      bulk: vi.fn(),
      bulkTags: vi.fn(),
      updateTags: vi.fn(),
      listTags: vi.fn().mockResolvedValue([]),
      listCategories: vi.fn().mockResolvedValue([]),
      createCategory: vi.fn(),
      updateCategory: vi.fn(),
      deleteCategory: vi.fn(),
      getAnalyticsSummary: vi.fn().mockResolvedValue(fixtures.summary),
      getAttendanceAnalytics: vi.fn().mockResolvedValue(fixtures.attendance),
      getConversionAnalytics: vi.fn().mockResolvedValue(fixtures.conversions),
      getApprovalFunnel: vi.fn().mockResolvedValue(fixtures.funnel),
      getEngagementHeatmap: vi.fn().mockResolvedValue(fixtures.heatmap)
    }
  }
})

vi.mock('../hooks/useWebSocket', () => {
  return {
    useWebSocket: () => ({
      readyState: 1,
      connectionAttempts: 0,
      subscribe: (handler: (event: MessageEvent) => void) => {
        subscribers.add(handler)
        hasActiveSubscribers = subscribers.size > 0
        return () => {
          subscribers.delete(handler)
          hasActiveSubscribers = subscribers.size > 0
        }
      },
      send: vi.fn(),
      close: vi.fn()
    })
  }
})

describe('Event analytics pipeline', () => {
  beforeEach(() => {
    vi.stubEnv('VITE_API_BASE_URL', 'http://localhost:4000')
    vi.stubEnv('VITE_WS_BASE_URL', 'ws://localhost:5000/ws')
  })

  afterEach(() => {
    vi.clearAllMocks()
    subscribers.clear()
    hasActiveSubscribers = false
    vi.unstubAllEnvs()
  })

  it('merges realtime WebSocket updates into analytics visuals', async () => {
    render(<EventManagement />)

    await waitFor(() =>
      expect(screen.getByTestId('analytics-total-events').textContent).toContain('42')
    )

    await waitFor(() => expect(hasActiveSubscribers).toBe(true))

    await act(async () => {
      emitWebSocketMessage({ type: 'summary', data: { totalEvents: 43 } })
    })

    await waitFor(() =>
      expect(screen.getByTestId('analytics-total-events').textContent).toContain('43')
    )

    await act(async () => {
      emitWebSocketMessage({ type: 'conversion', stage: 'Visiteurs', value: 520 })
    })

    await waitFor(() =>
      expect(screen.getByTestId('analytics-conversion-Visiteurs').textContent).toContain('520')
    )
  })

  it('captures reducer transformations snapshot', () => {
    const baseState: EventAnalyticsState = {
      summary: { ...fixtures.summary },
      attendance: [...fixtures.attendance],
      conversions: [...fixtures.conversions],
      funnel: [...fixtures.funnel],
      heatmap: [...fixtures.heatmap],
      updatedAt: '2024-01-01T00:00:00.000Z'
    }

    const actions: Parameters<typeof analyticsReducer>[1][] = [
      { type: 'PATCH_SUMMARY', payload: { totalEvents: 45, pendingApproval: 4 } },
      {
        type: 'UPSERT_ATTENDANCE_POINT',
        payload: {
          eventId: 'a',
          eventName: 'Forum Product',
          date: '2024-01-03',
          registrations: 155,
          attendance: 149
        }
      },
      { type: 'UPSERT_CONVERSION', payload: { stage: 'Présents', value: 150 } },
      { type: 'UPSERT_FUNNEL_STAGE', payload: { stage: 'Approuvés', count: 24 } },
      { type: 'UPSERT_HEATMAP_CELL', payload: { day: 'Mercredi', hour: 16, value: 90 } }
    ]

    const finalState = actions.reduce(analyticsReducer, baseState)
    const { updatedAt, ...stateWithoutTimestamp } = finalState

    expect(updatedAt).toEqual(expect.any(String))
    expect(stateWithoutTimestamp).toMatchSnapshot()
  })
})

