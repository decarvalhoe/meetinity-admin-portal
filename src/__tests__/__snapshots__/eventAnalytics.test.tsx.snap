// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Event analytics pipeline > captures reducer transformations snapshot 1`] = `
{
  "attendance": [
    {
      "eventId": "a",
      "eventName": "Forum Product",
      "series": [
        {
          "attendance": 110,
          "date": "2024-01-01",
          "registrations": 120,
        },
        {
          "attendance": 135,
          "date": "2024-01-02",
          "registrations": 140,
        },
        {
          "attendance": 149,
          "date": "2024-01-03",
          "registrations": 155,
        },
      ],
    },
  ],
  "conversions": [
    {
      "stage": "Visiteurs",
      "value": 420,
    },
    {
      "stage": "Inscrits",
      "value": 190,
    },
    {
      "stage": "Présents",
      "value": 150,
    },
  ],
  "funnel": [
    {
      "count": 40,
      "stage": "Soumis",
    },
    {
      "count": 28,
      "stage": "Vérification",
    },
    {
      "count": 24,
      "stage": "Approuvés",
    },
  ],
  "heatmap": [
    {
      "day": "Lundi",
      "hour": 9,
      "value": 30,
    },
    {
      "day": "Lundi",
      "hour": 10,
      "value": 50,
    },
    {
      "day": "Mardi",
      "hour": 14,
      "value": 70,
    },
    {
      "day": "Mercredi",
      "hour": 16,
      "value": 90,
    },
  ],
  "summary": {
    "averageAttendanceRate": 0.72,
    "conversionRate": 0.58,
    "pendingApproval": 4,
    "published": 30,
    "rejected": 6,
    "totalEvents": 45,
  },
}
`;
